src/main/java/com/android/volley/cronet/CronetHttpStack.java

```diff
@@ -0,0 +1,293 @@
+package com.android.volley.cronet;
+
+import android.content.Context;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+import com.android.volley.AuthFailureError;
+import com.android.volley.Header;
+import com.android.volley.Request;
+import com.android.volley.toolbox.AsyncHttpStack;
+import com.android.volley.toolbox.ByteArrayPool;
+import com.android.volley.toolbox.HttpResponse;
+import com.android.volley.toolbox.PoolingByteArrayOutputStream;
+import com.android.volley.toolbox.UrlRewriter;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.WritableByteChannel;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import org.chromium.net.CronetEngine;
+import org.chromium.net.CronetException;
+import org.chromium.net.UploadDataProvider;
+import org.chromium.net.UploadDataProviders;
+import org.chromium.net.UrlRequest;
+import org.chromium.net.UrlRequest.Callback;
+import org.chromium.net.UrlResponseInfo;
+
+/**
+ * A {@link AsyncHttpStack} that's based on Cronet's fully asynchronous API for network requests.
+ */
+public class CronetHttpStack extends AsyncHttpStack {
+
+    private final CronetEngine mCronetEngine;
+    private final Executor mCallbackExecutor;
+    private final ExecutorService mBlockingExecutor;
+    private final ByteArrayPool mPool;
+    private final UrlRewriter mUrlRewriter;
+
+    private CronetHttpStack(Builder builder) {
+        mCronetEngine = builder.mCronetEngine;
+        mCallbackExecutor = builder.mCallbackExecutor;
+        mBlockingExecutor = builder.mBlockingExecutor;
+        mPool = builder.mPool;
+        mUrlRewriter = builder.mUrlRewriter;
+    }
+
+    @Override
+    public void executeRequest(
+            final Request<?> request,
+            final Map<String, String> additionalHeaders,
+            final OnRequestComplete callback) {
+        final Callback urlCallback =
+                new Callback() {
+                    PoolingByteArrayOutputStream bytesReceived = null;
+                    WritableByteChannel receiveChannel = null;
+
+                    @Override
+                    public void onRedirectReceived(
+                            UrlRequest urlRequest, UrlResponseInfo urlResponseInfo, String s) {
```

Comment:
nit:  what is "s"? Assuming it has a more descriptive name in the Cronet documentation, could we use that name for clarity?